# Lesson 1: Introduction to Python

These notes are based on the sequence that concepts are introduced in the DataCamp lesson of the same name. They are most useful as a reference, after you've completed the DataCamp lesson. They are not intended to replace that lesson.

[View this lesson on DataCamp](https://learn.datacamp.com/courses/intro-to-python-for-data-science )

## Chapter 1
This chapter introduces the basics of Python - simple math and variable types

### Basic Operations

| Operator |  Operation                                                   |
|----------|--------------------------------------------------------------|
| `#`      |  add comment                                                 |
| `+`      |  addition                                                    |
| `-`      |  subtraction                                                 |
| `*`      |  multiplication                                              |
| `/`      |  division                                                    |
| `**`       |  exponentiation                                              |
| `%`        |  modulo; returns remainder of division (e.g., `18 % 7` equals `4`) |
| `=`        |  variable assignment                                         |
| `==`       |  equal to                                                    |
| `!=`       |  not equal to                                                |



### Variable Types

|                                          |                                                                      |
|------------------------------------------|----------------------------------------------------------------------|
| integer                                  |  number without fractional component (no decimal point)                                      |
| float                                    |  number with both integer and fractional part separated by a decimal |
| string  | text; enclose in `' '` or `" "`    |
| boolean                                  |  logical values; can only be `True` or `False`                       |
|list | Can contain many different types. Create with `[]`. Separate items with commas. |



### Functions
|           |                                       |
|-----------|---------------------------------------|
| `print()` |   outputs what's enclosed in `()`     |
| `type()`  |  gives the type of a defined variable |
| `int()`   |  converts a value into an integer     |
| `float()` |  converts a value into a float        |
| `str()`   |  converts a value into a string       |
| `bool()`  |  converts a value into a boolean      |
| `list()`  |  converts a value or values into a list |

## Chapter 2
This chapter covers everything to do with lists - creating lists, subsetting, and manipulation.

Lists are ordered sets of items. 

### Working with Lists
Create a list:

In [1]:
odd = [1, 3, 5, 7, 9]

First index (position) in list is 0, not 1:

In [2]:
odd[1]

3

In [3]:
odd[0]

1

Negative indexing works from the end of the list. 

In [4]:
odd[-1]

9

#### Slicing
Specify a range of indices inside square brackets: [inclusive:exclusive]

In [5]:
odd[0:5]

[1, 3, 5, 7, 9]

In [6]:
odd[1:3]

[3, 5]

In [7]:
odd[:]

[1, 3, 5, 7, 9]

In [8]:
odd[:3]

[1, 3, 5]

In [9]:
odd[3:]

[7, 9]

#### List Manipulation
Replacing list elements: subset list and assign a new value using =


In [10]:
odd[0] = 2
odd[0:5]

[2, 3, 5, 7, 9]

Extend a list, use + operator to add elements


In [11]:
odd + [11]

[2, 3, 5, 7, 9, 11]

In [12]:
odd  = [1, 3, 5, 7, 9]
even = [2, 4, 6, 8, 10]

all_numbers = odd + even
all_numbers

[1, 3, 5, 7, 9, 2, 4, 6, 8, 10]

Sort a list:

In [13]:
all_numbers.sort()
all_numbers

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Delete list items:

In [14]:
del(all_numbers[9])  # remember 0-indexing!
all_numbers

[1, 2, 3, 4, 5, 6, 7, 8, 9]

Copying a list

In [15]:
my_numbers = all_numbers.copy()

In [16]:
my_numbers

[1, 2, 3, 4, 5, 6, 7, 8, 9]

You can also use the `list` function to make a list from another list:

In [17]:
some_numbers = list(all_numbers)

In [18]:
some_numbers

[1, 2, 3, 4, 5, 6, 7, 8, 9]

Be careful not to do this. This "points" the new list, `my_numbers_nono`, at the original list, `all_numbers`

In [19]:
my_numbers_nono = all_numbers
my_numbers_nono

[1, 2, 3, 4, 5, 6, 7, 8, 9]

So when you modify `all_numbers`, you also change `my_numbers_nono`

In [20]:
del(all_numbers[8])
my_numbers_nono

[1, 2, 3, 4, 5, 6, 7, 8]

## Chapter 3

 This chapter introduces a variety of functions and methods, as well as Python packages

In [21]:
my_list = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

### Functions

#### Max/min values in list

In [22]:
max(my_list)

10

In [23]:
min(my_list)

1

#### Rounding

In [24]:
pi = 3.14159265359
round(pi, 2)

3.14

#### Length of list

In [25]:
len(my_list)

10

#### Help

In [26]:
?round

Signature: round(number, ndigits=None)
Docstring:
Round a number to a given precision in decimal digits.

The return value is an integer if ndigits is omitted or None.  Otherwise
the return value has the same type as the number.  ndigits may be negative.
Type:      builtin_function_or_method


#### Sorting

In [27]:
going_up = sorted(my_list)
going_up

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [28]:
going_down = sorted(going_up, reverse=True)
going_down

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

### Methods

Methods are functions that are applied to a variable by adding a dot then the method name (and parentheses) to the end of the variable name.

| Method | What it does |
|--------|--------------|
|`.index()` | for a list, gives index of the input value |
|`.capitalize()` | returns string with the first letter capitalized |
|`.upper()` | capitalizes all letters in a string |
|`.count()` | for `string`: counts how many of the input are in a string |
|`.count()` | for `list`: counts the number of times an element appears in a list |
|`.append()` | adds an element to the list it is called on |
|`.remove()` | removes the first element of a list that matches the input |
|`.reverse()` | reverses the order of the elements in the list its called on |
|`.round()` | rounds a float to the number of decimal places specified in the parentheses |

In [29]:
going_down.index(10)

0

In [30]:
going_down.append(0)
going_down

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

### Packages

Import a package (sometimes called a library), including all its functions

In [31]:
import matplotlib

Import a package and give it a shorthand name:

In [32]:
import numpy as np

## Chapter 4

This chapter introduces the `numpy` package (by convention, imported as `np`), focusing on the numpy array data type.

### NumPy arrays

NumPy arrays are similar to lists: ordered sets of values. However, they behave differently, and can have multiple dimensons (see below), so NumPy arrays are often a better choice for structured data.

You can **create** an np array manually by putting a list inside the `np.array()` function:

In [33]:
np_odds = np.array([1, 3, 5, 7, 9])
np_odds

array([1, 3, 5, 7, 9])

Or by specifying the name of a list you previously created:

In [34]:
evens = [2, 4, 6, 8, 10]
np_evens = np.array(evens)
np_evens

array([ 2,  4,  6,  8, 10])

**Note**: If you just type the name of an array, you'll see the the contents of the array, embedded inside `array()`, as shown above. If you `print()` the array, you'll see only its contents:

In [35]:
print(np_evens)

[ 2  4  6  8 10]


Either way above is acceptable, but remember that in Jupyter, you need to use `print()` to see the output of anything other than the last returned value:

In [36]:
np_evens
np_odds

array([1, 3, 5, 7, 9])

In [37]:
print(np_evens)
np_odds

[ 2  4  6  8 10]


array([1, 3, 5, 7, 9])

NumPy arrays can only contain values of one type. Values are converted to a different type to enforce this:

In [38]:
np.array([1, 'two', False])

array(['1', 'two', 'False'], dtype='<U21')

Above, the int `1` and the Boolean `False` were converted to a string, because the string 'two' couldn't be converted to any other type.

Below, we only have ints and Booleans. Booleans are converted to ints:

In [39]:
np.array([1, 2, False, True])

array([1, 2, 0, 1])

Unlike lists, operations are applied to np arrays element-wise (i.e., applies operation to each element in the array).

So for a list, this is probably *not* your intended outcome:

In [40]:
print(evens * 2)

[2, 4, 6, 8, 10, 2, 4, 6, 8, 10]


But for a NumPy array:

In [41]:
print(np_evens * 2)

[ 4  8 12 16 20]


### 2D arrays

A NumPy array with two dimensions (rows and columns).

Can create from a regular python list of lists (note the two lists are inside another list/set of square brackets):

In [42]:
my_2d_array = np.array([[0, 1, 2], [3, 4, 5]])

In [43]:
my_2d_array

array([[0, 1, 2],
       [3, 4, 5]])

#### Indexing 2D Arrays:
Specify in format [*rows*, *columns*]. As with lists, `:` selects entire row/column, and you can use negative indexing

In [44]:
my_2d_array[0, :]

array([0, 1, 2])

In [45]:
my_2d_array[1, :]

array([3, 4, 5])

Column only:

In [46]:
my_2d_array[:, 0]

array([0, 3])

Last column:

In [47]:
my_2d_array[:, -1]

array([2, 5])

Select single value by specifying a particular row and column:

In [48]:
my_2d_array[1, 1]

4

`.shape` *attribute* gives shape (rows, cols) of 2D numpy array:

In [49]:
my_2d_array.shape

(2, 3)

### NumPy Array Functions

In [50]:
np.random.rand(10)

array([5.59828405e-01, 4.60389224e-01, 2.13615839e-01, 5.03037469e-01,
       1.75239467e-01, 8.06776497e-05, 5.82495763e-01, 7.78231303e-01,
       5.00786071e-01, 9.63882007e-02])

In [51]:
# Some random numbers:
my_data = np.random.rand(10)
print(my_data)

[0.93528333 0.21361937 0.5551476  0.59618613 0.20647105 0.6735143
 0.66893563 0.9022425  0.55059742 0.13539459]


#### Mean

In [52]:
np.mean(my_data)

0.5437391931458195

#### Median

In [53]:
np.median(my_data)

0.5756668682826799

#### Standard Deviation

In [54]:
np.std(my_data)

0.2658475578053544

If you don't like long numbers, you can apply the .round() method to the output:

In [55]:
std_dev = np.std(my_data).round(3)
print(std_dev)

0.266


#### Correlation 

Between two arrays:

In [56]:
# Final grades
grades = np.array([67, 74, 44, 92, 88])

# hours per week spent working on course
hours = np.array([5.2, 7, 2.0, 9.25, 7.5])

Does more hours spent on the course translate into a better grade?

In [57]:
np.corrcoef(grades, hours)

array([[1.        , 0.97877197],
       [0.97877197, 1.        ]])

Yes! The correlation is positive and strong (approximately .979)